File: src/modules/role/services/role-delete.service.ts
import { Injectable, NotFoundException } from '@nestjs/common';
import { PrismaService } from 'src/infra/database/prisma.service';
import { RoleEntity } from '../domain/entities/role.entity';
import { UserRequest } from 'src/shared/types/user-request';
import { roleDefaultValidator } from '../validators/role-default.validator';
import { roleReferenceValidator } from '../validators/role-reference.validator';
import { roleDeleteValidator } from '../validators/role-delete.validator';

@Injectable()
export class RoleDeleteService {
  constructor(private prisma: PrismaService) {}
  async execute(id: number, userRequest: UserRequest): Promise<RoleEntity> {
    const role = await this.prisma.role.findFirst({
      where: {
        id: id,
        organizationId: userRequest?.organizationId ?? undefined,
        companyId: userRequest?.companyId ?? undefined,
        deletedAt: null,
      },
    });

    if (!role) throw new NotFoundException('Perfil não encontrado!');

    await roleDefaultValidator(role);

    await roleReferenceValidator(role.reference, userRequest);

    await roleDeleteValidator(id);

    const updatedRole = await this.prisma.role.update({
      data: {
        deletedAt: new Date(),
      },
      where: {
        id: id,
      },
      include: {
        organization: true,
        company: true,
      },
    });

    await this.prisma.roleAbility.deleteMany({
      where: { roleId: updatedRole.id },
    });

    return new RoleEntity(updatedRole);
  }
}

File: src/modules/role/services/role_delete_service.py
from django.db.models import Q
from django.utils import timezone
from src.shared.types.user_request import UserRequest
from src.modules.role.models import Role, RoleAbility
from src.shared.exceptions.not_found_exception import NotFoundException
from src.modules.role.validators.role_delete_validator import (
    role_delete_validator,
)
from src.modules.role.validators.role_default_validator import (
    role_default_validator,
)
from src.modules.role.validators.role_reference_validator import (
    role_reference_validator,
)


class RoleDeleteService:
    @classmethod
    def execute(self, id: int, user_request: UserRequest):

        role = Role.objects.filter(
            Q(id=id)
            & Q(deleted_at=None)
            & (
                Q(organization__id=user_request.organization.id)
                if user_request.organization
                else Q()
            )
            & (Q(company__id=user_request.company.id) if user_request.company else Q())
        ).first()

        if role is None:
            raise NotFoundException("Perfil não encontrado!")

        role_default_validator(role)

        role_reference_validator(role.reference, user_request)

        role_delete_validator(id)

        role.deleted_at = timezone.now()
        role.save()

        RoleAbility.objects.filter(role__id=id).delete()

        return role
